src/data/fintype/card_embedding.lean

```diff
@@ -0,0 +1,94 @@
+/-
+Copyright (c) 2021 Eric Rodriguez. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Rodriguez
+-/
+import data.fintype.card
+import data.equiv.fin
+import data.nat.factorial
+import data.equiv.embedding
+
+/-!
+# Birthday Problem
+
+This file establishes the cardinality of `α ↪ β` in full generality.
+-/
+
+local notation `|` x `|` := finset.card x
+local notation `‖` x `‖` := fintype.card x
+
+open_locale nat
+
+namespace fintype
+
+-- the [fintype α] argument may seem redundant; it isn't!
+lemma card_embedding_of_unique
+{α β : Type*} [unique α] [fintype α] [fintype β] [decidable_eq α] [decidable_eq β]:
+‖α ↪ β‖ = ‖β‖ := card_congr equiv.unique_embedding_equiv_result
+
+private lemma card_embedding_aux (n : ℕ) (β) [fintype β] [decidable_eq β] (h : n ≤ ‖β‖) :
+  ‖fin n ↪ β‖ = nat.desc_fac (‖β‖ - n) n :=
+begin
+  induction n with n hn,
+  { nontriviality (fin 0 ↪ β),
+    rw [nat.desc_fac_zero, fintype.card_eq_one_iff],
+    refine ⟨nonempty.some nontrivial.to_nonempty, λ x, function.embedding.ext fin.elim0⟩ },
+
+  rw [nat.succ_eq_add_one, ←card_congr (equiv.embedding_congr fin_sum_fin_equiv (equiv.refl β))],
+  rw card_congr equiv.sum_embedding_equiv_sigma_embedding_restricted,
+  all_goals { try { apply_instance } },
+
+  have : ∀ (f : fin n ↪ β), ‖fin 1 ↪ ↥((set.range f).compl)‖ = ‖β‖ - n,
+  {
+    intro f,
+    rw card_embedding_of_unique; try {apply_instance},
```

Comment:
I think this is abuse of `;` and `try`. Just use `apply_instance` where it is needed.

Reply:
ahh, I can actually remove this now; originally `card_embedding_of_unique` (or the `card_congr` lemma) just refused to use inference, and actually instead set them as subgoals; now that it's packaged up, inference ends up working fine